src/bin/jcli_app/utils/output_format.rs

```diff
@@ -0,0 +1,83 @@
+use gtmpl::Value as GtmplValue;
+use jcli_app::utils::error::CustomErrorFiller;
+use serde::Serialize;
+use serde_json::{Map as JsonMap, Number as JsonNumber, Value as JsonValue};
+use std::fmt::{self, Display, Formatter};
+use structopt::StructOpt;
+
+#[derive(StructOpt)]
+pub struct OutputFormat {
+    /// Format of output data. Possible values: json, yaml.
+    /// Any other value is treated as a custom format using values from output data structure.
+    /// Syntax is Go text template: https://golang.org/pkg/text/template/.
+    #[structopt(long = "output-format", default_value = "yaml")]
+    format: String,
+}
+
+custom_error! { pub Error
+    YamlFormattingFailed { source: serde_yaml::Error } = "failed to format output as YAML",
+    JsonFormattingFailed { source: serde_json::Error } = "failed to format output as JSON",
+    JsonValueCreateFailed { source: serde_json::Error, filler: CustomErrorFiller }
+        = "failed to create intermediate formatting data",
+    CustomFormattingFailed { source: GtmplError } = "failed to format output as custom format",
+}
+
+#[derive(Debug)]
+pub struct GtmplError(String);
+
+impl Display for GtmplError {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), fmt::Error> {
+        self.0.fmt(f)
+    }
+}
+
+impl std::error::Error for GtmplError {}
+
+impl OutputFormat {
+    pub fn format(&self, data: impl Serialize) -> Result<String, Error> {
+        Ok(match self.format.trim().to_ascii_lowercase().as_str() {
+            "yaml" => serde_yaml::to_string(&data)?,
+            "json" => serde_json::to_string_pretty(&data)?,
+            _ => {
+                let value = serde_json::to_value(data)
+                    .map(json_value_to_gtmpl)
+                    .map_err(|source| Error::JsonValueCreateFailed {
+                        source,
+                        filler: CustomErrorFiller,
```

Comment:
:confused:

Reply:
This `filler` pattern is the best I could come up while fighting with `custom_error` macro. Without it there is a conflict of `From`s for variant with same `source`. And without `source` `Error::cause` doesn't work. Author of this macro assumed that `from` and `cause` functionalities should be synonymous.

Reply:
You are talking about `cause` as if it's not deprecated. Is `custom_error` unaware of `Error::source`?

`custom_error` is only a convenience. If it does not cut it, there's always the good old way of manually defining `Error`, `Display`, etc.

I feel a growing urge to try `snafu` for error definition.

Reply:
Right, I meant `Error::source`.

I think that this spot of ugliness is still a decent trade-off for not having to maintaining `source` and `Display` for every variant. Definitely migrating to a better tool is a great idea, `custom_error` isn't that good. Just not in this PR.